Lib9c/Model/State/ShopState.cs

```diff
@@ -9,126 +10,236 @@
+                    {
+                        _agentProducts.Remove(sellerAgentAddress);
+                    }
                 }
+            }
 
-                outPair = new KeyValuePair<Address, ShopItem>(sellerAgentAddress, shopItem);
-                return true;
+            var itemSubType = targetShopItem.ItemUsable.ItemSubType;
+            if (_itemSubTypeProducts.ContainsKey(itemSubType))
+            {
+                var shopItems = _itemSubTypeProducts[itemSubType];
+                shopItems.Remove(productId);
+                if (shopItems.Count == 0)
+                {
+                    _itemSubTypeProducts.Remove(itemSubType);
+                }
             }
 
-            return false;
+            unregisteredItem = targetShopItem;
+            return true;
         }
 
-        public bool TryUnregister(Address sellerAgentAddress,
-            Guid productId, out ShopItem outUnregisteredItem)
+        #endregion
+
+        public bool TryGet(
+            Address sellerAgentAddress,
+            Guid productId,
+            out ShopItem shopItem)
         {
-            if (!TryGet(sellerAgentAddress, productId, out var outPair))
+            if (!_agentProducts.ContainsKey(sellerAgentAddress))
+            {
+                shopItem = null;
+                return false;
+            }
+
+            var shopItems = _agentProducts[sellerAgentAddress];
+            if (shopItems.All(item => item != productId))
             {
-                outUnregisteredItem = null;
+                shopItem = null;
                 return false;
             }
 
-            AgentProducts[outPair.Key].Remove(outPair.Value);
+            if (!_products.ContainsKey(productId))
+            {
+                shopItem = null;
+                return false;
+            }
 
-            outUnregisteredItem = outPair.Value;
+            shopItem = _products[productId];
             return true;
         }
+    }
+
+    [Serializable]
+    public class ShopStateAlreadyContainsException : Exception
+    {
+        public ShopStateAlreadyContainsException(string message) : base(message)
+        {
+        }
 
-        public override IValue Serialize() =>
-            new Dictionary(new Dictionary<IKey, IValue>
-            {
-                [(Text)"agentProducts"] = new Dictionary(
-                    AgentProducts.Select(kv =>
-                        new KeyValuePair<IKey, IValue>(
-                            (Binary)kv.Key.Serialize(),
-                            new List(kv.Value.Select(i => i.Serialize()))
-                        )
-                    )
-                )
-            }.Union((Dictionary)base.Serialize()));
+        protected ShopStateAlreadyContainsException(
+            SerializationInfo info,
+            StreamingContext context) : base(info, context)
+        {
+        }
+    }
+
+    [Serializable]
+    public class FailedToUnregisterInShopStateException : Exception
```

Comment:
Can you please move this exception to new file? It's not necessary, but easy to look up in IDEs :)